### ISSUE TITLE
Retry to start workflow if failed for RequestLimitExceeded or InsufficientInstanceCapacity after some time

### ISSUE THREAD

**Issue by MisterJimson** [runs-on/runs-on#28]
Currently when a runs-on error occurs and an email is sent, the workflow stays queued forever. The user must cancel and retry the workflow.

It would be nice if there was a time period where runs-on would try again, perhaps with some backoff logic. The default could be 1 hour -> 2 ->  4 -> 8 -> give up.

The other alternative approach would be to fail the workflow, so it's clear in Github there was an error instead of having the workflow queued forever.

**Comment by crohr** [comment]
Regarding `RequestLimitExceeded`, I now have a proper benchmark with a few hundred runners started at once to reproduce the issue, and I believe this will be fixed for good in the next version.

But yes for `InsufficientInstanceCapacity`, the workflow should be retried at a later time, or possibly automatically fall back to an alternative available instance type, even if it might be slower? What do you think?

**Comment by MisterJimson** [comment]
I would love for it to fallback to a different runner!

RequestLimitExceeded we were running into with a small workflow, it was because we hand't used EC2 on this account so the limits were low. They are fine now, but retrying it would smooth onboarding for other new folks.

**Comment by crohr** [comment]
@MisterJimson was it a rate limit for the number of instances, or limit for API call such as RunInstances?

**Comment by MisterJimson** [comment]
The RequestLimitExceeded was for the API call I believe. Here is the full message.

```
Error: Unable to start EC2 instance with the following configuration: {"imageSpec":{"platform":"linux","arch":"x64","name":"runs-on-ubuntu22-full-x64-*","owner":"135269210855"},"runnerSpec":{"cpu":2,"family":["m7a","m7g"],"pricing":[0.001932,0.000783],"on_demand_price_per_min":"0.0022","spot_price_per_min":"0.0010","github_price_per_min":0.008,"github_ratio":"8"}}. The error was: RequestLimitExceeded: Request limit exceeded.
 ```
 
 When it comes to the number of instances, we did hit `VcpuLimitExceeded` while waiting for quota increases from AWS.

**Comment by crohr** [comment]
Thanks, hopefully this one is fixed with v1.7.1, coming next week. Or if you like living on the edge, you can already try it with this CF template: https://runs-on.s3.eu-west-1.amazonaws.com/cloudformation/template-dev.yaml

**Comment by crohr** [comment]
Also if you're interested in joining a Slack channel for RunsOn, send me an email at cyril@runs-on.com

**Comment by pvande** [comment]
I'll also chime in on this. I've been running into `InsufficientInstanceCapacity` with unfortunate regularity lately. (I've just shifted to a new AZ, so 🤞…)

For our concerns, we're not particular about _which_ AZ it runs in, or all that particular about the runner's size (we do have some minimum thresholds, though). What's more important for our needs is that the jobs get run promptly — our jobs tend to run in minutes, but we run a quite a few of them.

> You can currently get m7a.large capacity by not specifying an Availability Zone in your request or choosing us-east-1b, us-east-1c, us-east-1d, us-east-1f.

This message from AWS is also intriguing, though not specifying an AZ doesn't appear to be something we can currently configure.

**Comment by crohr** [comment]
@pvande if you're not too picky about instance types, specify `family=m7i+c6a` for instance, and you should not have any issues in starting an instance.

I could make RunsOn multi-az, but this would increase the time for booting instances since you would have to try multiple AZs, and then possibly multiple instance types per AZ. The solution would be to use an EC2 fleet request, but unfortunately it's not as flexible as I'd like it to be (especially for specifying custom images). I'm currently experimenting with it though.

I have also updated the [troubleshooting guide](https://runs-on.com/guides/troubleshoot/) with further details on what you can do in case you hit this issue.

**Comment by crohr** [comment]
Also relevant, I've fixed another rate limit issue due to the default retry behaviour of the official AWS SDK, which forcefully depletes the RunInstances quota on client errors (which somehow `InsufficientInstanceCapacity` is!): https://github.com/runs-on/runs-on/pull/29

Will publish a new release shortly.

**Comment by crohr** [comment]
I've released v1.7.3, which should fix the `RequestLimitExceeded` for this case. Also added `m7i` as an additional instance family for x64, which should help with the current short supply of m7a/c7a instances. They are a bit slower, but also cheaper :)

I'm going to close this issue for now, thanks everyone!
